client/src/reducers/genesets.ts

```diff
@@ -63,24 +63,37 @@ const GeneSets = (
       const genesets = new Map();
 
       for (const gsData of genesetsData) {
-        const genes = new Map();
-        for (const gene of gsData.genes) {
-          genes.set(gene.gene_symbol, {
-            geneSymbol: gene.gene_symbol,
-            geneDescription: gene?.gene_description ?? "",
-          });
+        if (
+          gsData &&
+          typeof gsData === "object" &&
+          !Array.isArray(gsData) &&
+          typeof gsData.geneset_name === "string"
+        ) {
```

Comment:
Front-end counterpart to the TileDB changes: we now normalize the server payload and track `lastTid`. That TID is incremented client-side before issuing a PUT so we fail fast if the backend rejects stale writes.